### DISCUSSION TITLE
There are potential problems with Array2DHashSet and FlexibleHashMap.

### DISCUSSION THREAD

**Discussion by yososs** [antlr/antlr4#3712] [General]
Incorrect threshold when initial container size is specified in the constructor.
The clear method ignores the container size specified in the constructor. Also, the threshold is not recalculated.

* https://github.com/antlr/antlr4/blob/master/runtime/Java/src/org/antlr/v4/runtime/misc/Array2DHashSet.java
* https://github.com/antlr/antlr4/blob/master/runtime/Java/src/org/antlr/v4/runtime/misc/FlexibleHashMap.java

Currently, these are not causing any real problems because of the following assumptions.

* Using an initial container size of 16.
* They are not (presumably) using the clear method.

In class member declarations, the values of threshold are as follows

``` java
public static final int INITAL_CAPACITY = 16; // must be power of 2

protected int threshold = (int)Math.floor(INITAL_CAPACITY * LOAD_FACTOR); // when to expand

```
Thus, even if a large value is specified for initialCapacity in the Array2DHashSet constructor, the first expansion is done at the threshold value of 16*0.75.

The implementation of clear also ignores the constructor parameter and is considered incorrect.

The code to correct this problem is attached.

* [patch.zip](https://github.com/antlr/antlr4/files/8699949/patch.zip)

**Comment by parrt** [comment]
yep, makes sense. code is wrong in antlr; but didn't affect anything. whew. Making a PR.  https://github.com/antlr/antlr4/pull/3726

**Reply by KvanTTT** [reply]
Probably it should be converted to an issue and closed.
